src/common/ceph_time.h

```diff
@@ -452,62 +459,30 @@ namespace ceph {
   // time_point + duration to assert on overflow, but I don't think we
   // should.
 
-} // namespace ceph
 
-// We need these definitions to be able to hande ::encode/::decode on
-// time points.
-
-template<typename Clock, typename Duration>
-void encode(const std::chrono::time_point<Clock, Duration>& t,
-	    ceph::bufferlist &bl) {
-  auto ts = Clock::to_timespec(t);
-  // A 32 bit count of seconds causes me vast unhappiness.
-  uint32_t s = ts.tv_sec;
-  uint32_t ns = ts.tv_nsec;
-  ::encode(s, bl);
-  ::encode(ns, bl);
+inline timespan abs(signedspan z) {
+  return z > signedspan::zero() ?
+    std::chrono::duration_cast<timespan>(z) :
+    timespan(-z.count());
 }
-
-template<typename Clock, typename Duration>
-void decode(std::chrono::time_point<Clock, Duration>& t,
-	    bufferlist::iterator& p) {
-  uint32_t s;
-  uint32_t ns;
-  ::decode(s, p);
-  ::decode(ns, p);
-  struct timespec ts = {
-    static_cast<time_t>(s),
-    static_cast<long int>(ns)};
-
-  t = Clock::from_timespec(ts);
+inline timespan to_timespan(signedspan z) {
+  ceph_assert(z >= signedspan::zero());
+  return std::chrono::duration_cast<timespan>(z);
 }
 
-// C++ Overload Resolution requires that our encode/decode functions
-// be defined in the same namespace as the type. So we need this
-// to handle things like ::encode(std::vector<ceph::real_time // > >)
+// detects presence of Clock::to_timespec() and from_timespec()
+template <typename Clock, typename = std::void_t<>>
```

Comment:
@cbodley I'm guessing `std::void_t` is a no-no in luminous?

Reply:
yeah, afraid so. i think the best way forward (backward?) now is to drop those extra commits, and just add a single utime_t overload for coarse_real_time. i've pushed a copy of this branch to https://github.com/cbodley/ceph/commits/wip-35980 with a new commit https://github.com/cbodley/ceph/commit/df29bb3d2d8ae302e8ed1227421885ba2a5aea88 that does that